MarkEditMac/Sources/Scripting/EditorDocument+Scripting.swift

```diff
@@ -0,0 +1,191 @@
+//
+//  EditorDocument+Scripting.swift
+//  MarkEdit
+//
+//  Created by Stephen Kaplan on 4/2/25.
+//
+
+import MarkEditKit
+
+extension EditorDocument {
+  /// Raw markdown string of the document.
+  @objc var scriptingSource: String {
+    get {
+      return stringValue
+    }
+    set {
+      if newValue != stringValue {
+        guard let targetEditor = scriptingTargetEditor(for: nil) else { return }
+        stringValue = newValue
+        targetEditor.bridge.core.replaceText(text: newValue, granularity: .wholeDocument)
+      }
+    }
+  }
+
+  /// Rich text of the document.
+  @objc var scriptingRichText: NSTextStorage? {
+      let markdownOptions = AttributedString.MarkdownParsingOptions(
+        allowsExtendedAttributes: true,
+        interpretedSyntax: .full,
+        failurePolicy: .returnPartiallyParsedIfPossible,
+        appliesSourcePositionAttributes: true
+      )
+
+      // Process lines separately to ensure they stay distinct
+      let newlineString = NSAttributedString(string: "\n")
+      let attributedSource = scriptingSource.components(separatedBy: .newlines)
+      .map { component in
+        do {
+          return try NSMutableAttributedString(markdown: component, options: markdownOptions)
+        } catch {
+          return NSMutableAttributedString(string: component)
+        }
+      }
+      .reduce(into: NSMutableAttributedString()) { result, element in
+        element.append(newlineString)
+        result.append(element)
+      }
+
+      return NSTextStorage(attributedString: attributedSource)
+  }
+
+  @objc var scriptingSelectedText: String {
+    get {
+      guard let command = currentScriptCommand(),
+            let targetEditor = scriptingTargetEditor(for: command) else {
+        return ""
+      }
+
+      command.suspendExecution()
+      Task {
+        let text = try await targetEditor.bridge.selection.getText()
+        command.resumeExecution(withResult: NSAppleEventDescriptor(with: text))
+      }
+      return self.stringValue
+    }
+    set {
+      guard let targetEditor = scriptingTargetEditor(for: nil) else { return }
+      Task {
+        targetEditor.bridge.core.replaceText(text: newValue, granularity: .selection)
+      }
+    }
+  }
+
+  /// Executes a user-provided JS script in the document's webview.
+  @objc func scriptingHandleEvaluateCommand(_ command: NSScriptCommand) -> Any? {
+    guard let inputString = command.evaluatedArguments?["script"] as? String else {
+      ScriptingError.missingArgument("script").applyToCommand(command)
+      return nil
+    }
+
+    guard let targetEditor = scriptingTargetEditor(for: command) else {
+      return nil
+    }
+
+    // Preserve newlines for better error reporting
+    let jsText = inputString.replacingOccurrences(of: "\\n", with: "\\\\n")
```

Comment:
@SKaplanOfficial this quite interesting to me, does this really work?

If we are replacing `\n` to `\\n` in the scripts, should it be `inputString.replacingOccurrences(of: "\n", with: "\\n")`?

I don't quite get why we need double backslashes, and why we need to replace them at all.